GoogleForCreators/web-stories-wp
# Dashboard: Create story from template

## Summary
Adds ability to create a story from a template

## Relevant Technical Choices
- Adds new action in useStoryApi that takes template and retrieves rest of story info from `getStoryPropsToSave` and then adds that template as a story
- Exposed template version because without it the new story-template breaks in the editor

(Some clean up)
- New shared view component for template grids (`templateGridView`) to be used in explore templates and template details. The shared storyGrid was just getting so many props and having to handle dueling functionality. Separating this cleaned up the grid components themselves and made the view structure easier. 
- Updating unit tests that rely on `grid-item` to have unique test ids and use regex to grab them all 
- New storybook for templateGridView 
- Move view related storybook content into dashboard/views instead of in components to keep a clear break of content

## User-facing changes
- Ability to create a story from a template. Now when you click 'use template' on 'Explore templates' you get redirected to the editor with an auto-draft copy of that template as a story.

## Testing Instructions
- Regression test explore templates and detail templates, make sure grids still work. 
- See new storybook set up with dashboard/views and new storybook for templateGridView
- Click **'use template'** on any template in the dashboard, you'll see a `post` to `wp-json/wp/v2/web-story?_locale=user` that should return a `201 created` on success. When successful it will redirect to the editor with the new ID created for the auto draft. If you save the draft or publish this template has become a story and is visible in dashboard stories. if you do not save it, it is not available in my stories on dashboard. 
- if you fudge the API and make it error you'll get an error object back that sets in the reducer. We're not handling error states yet in the UI but I did update the reducer to bring back a message instead of a boolean. 
- other places this functionality exists aside from the 'use template' button in the explore templates grid on hover: detail template view (seen when you click 'view') on hover of a template (middle button) x 2 spots. 1) the 'use template' blue CTA in the nav section. 2) the grid of related templates on hover (just like in explore templates) 

---



addresses #2007
